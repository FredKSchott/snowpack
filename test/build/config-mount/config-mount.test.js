const fs = require('fs');
const path = require('path');
const glob = require('glob');
const cheerio = require('cheerio');
const {fileLines} = require('../../test-utils');

function generateContentsMap(dir) {
  const contentMap = {};
  const allFiles = glob.sync(dir + '**/*', {nodir: true});
  allFiles.sort((a, b) => a.localeCompare(b, {numeric: true}));
  allFiles.forEach((file) => {
    contentMap[path.relative(dir, file)] = fs.readFileSync(file, 'utf-8');
  });
  return contentMap;
}

describe('config: mount', () => {
  describe('basic', () => {
    const tests = [
      {
        name: 'allows direct mappings',
        given: 'a',
        expect: 'a',
      },
      {
        name: 'allows renamed mappings',
        given: 'src/b',
        expect: 'new-b',
      },
      {
        name: 'allows renamed mappings 2',
        given: 'src/c',
        expect: 'deep/c',
      },
      {
        name: 'allows ending file slash',
        given: 'src/d',
        expect: 'bad/d', // this mount key has a trailing slash in config, but it should still appear at this path
      },
      {
        name: 'allows renamed mappings 3',
        given: 'src/e/f',
        expect: 'e',
      },
      {
        name: 'static + resolve',
        given: 'src/j',
        expect: 'j',
      },
    ];

    tests.forEach((t) => {
      it(t.name, () => {
        const given = generateContentsMap(path.join(__dirname, t.given));
        const expected = generateContentsMap(path.join(__dirname, 'build', t.expect));
        expect(given).toEqual(expected);
      });
    });
  });

  describe('advanced', () => {
    it('url', () => {
      const cwd = path.join(__dirname, 'build', 'new-g');
      const distJS = fileLines(path.join(cwd, 'index.js'));
      const $ = cheerio.load(fs.readFileSync(path.join(cwd, 'main.html'), 'utf-8'));

      expect(distJS[1]).toBe(`import "./dep.js";`); // formatter ran
      expect($('script[type="module"]').attr('src')).toBe('/_dist_/index.js'); // JS resolved
    });

    it('static', () => {
      const cwd = path.join(__dirname, 'build', 'h');
      const $ = cheerio.load(fs.readFileSync(path.join(cwd, 'main.html'), 'utf-8'));

      expect($('script[type="module"]').attr('src')).toBe('/_dist_/index.js'); // JS resolved
    });

    it('resolve: false', () => {
      const cwd = path.join(__dirname, 'build', 'i');
      const distJS = fileLines(path.join(cwd, 'index.js'));

      expect(distJS[1]).toBe(`import "./dep";`); // JS not resolved
    });
  });
});
