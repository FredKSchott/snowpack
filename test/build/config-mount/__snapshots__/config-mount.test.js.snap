// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config: mount basic allows direct mappings: allows direct mappings 1`] = `
Object {
  "index.js": "console.log('a');",
}
`;

exports[`config: mount basic allows ending file slash: allows ending file slash 1`] = `
Object {
  "index.js": "console.log('d');",
}
`;

exports[`config: mount basic allows renamed mappings 2: allows renamed mappings 2 1`] = `
Object {
  "index.js": "console.log('c');",
}
`;

exports[`config: mount basic allows renamed mappings 3: allows renamed mappings 3 1`] = `
Object {
  "index.js": "console.log('e/f');",
}
`;

exports[`config: mount basic allows renamed mappings: allows renamed mappings 1`] = `
Object {
  "index.js": "console.log('b');",
}
`;

exports[`config: mount basic static + resolve: static + resolve 1`] = `
Object {
  "dep.js": "console.log('dep');",
  "index.jsx.js": "// TEST NOTE: since resolve=false, expected to be broken import
import './dep';
console.log('i');
",
  "main.html": "<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"Web site created using create-snowpack-app\\" />
    <title>Snowpack App</title>
  </head>
  <body>
    <!-- TEST NOTE: since resolve=false, expected to be broken import -->
    <script type=\\"module\\" src=\\"%PUBLIC_URL%/_dist_/index.js\\"></script>
    <script type=\\"module\\">
        // TEST NOTE: since resolve=false, expected to be broken import
        import './dep';
        console.log('main');
    </script>
  </body>
</html>
",
}
`;
